### PR TITLE
analyzer for changes of environment variables in meshConfig.defaultConfig

### LINKED ISSUE #32602: analyzer for changes in meshConfig.defaultConfig
Feature request:

If I install Istio with `--set meshConfig.defaultConfig.tracing.zipkin.address=jaeger-collector.istio-system.svc.cluster.local:9411`, deploy pods, then re-install Istio with `--set meshConfig.defaultConfig.tracing.zipkin.address=jaeger-collector.istio-system.svc.cluster.local:14268`, the sidecars will have stale data.

They should be restarted, but the only value analysis currently checks for is the image name of the sidecar.

### PR DESCRIPTION
**Pull request by zhlsunshine**
**Please provide a description of this PR:**

This PR fixed [issue#32602](https://github.com/istio/istio/issues/32602) and [issue#34439](https://github.com/istio/istio/issues/34439), the former is for re-deploying Istio, the latter is for changing the environment variables in meshConfig.defaultConfig,  such as  `PROXY_CONFIG` and any other env variable in `proxyMetadata`

### CODE REVIEW

**Comment by zhlsunshine** [comment]
Here are the output steps below:
1. Install istio by `istioctl install --set profile=demo -y --set meshConfig.defaultConfig.tracing.zipkin.address=jaeger-collector.istio-system.svc.cluster.local:9411`
2. Deploy httpbin and sleep by `kubectl apply -f samples/httpbin/httpbin.yaml` and `kubectl apply -f samples/sleep/sleep.yaml`
3. Check pod's sidecar data by `kubectl describe pods productpage-v1-xxxxxx` in environment `PROXY_CONFIG`  of container `istio-proxy`
4. Re-install istio by `istioctl install --set profile=demo -y --set meshConfig.defaultConfig.tracing.zipkin.address=jaeger-collector.istio-system.svc.cluster.local:14268` or change the value directly in configmap `istio` under the namespace of `istio-system`
5. Analyze via `istioctl analyze -A` and the result looks like below:
```
Warning [IST0149] (Pod httpbin-74fb669cc6-9tcrx.default) The environment variable configured with the running proxies on pods does not match the environment variable defined in the mesh configmap (dismatch environment variable in pod is {"tracing":{"zipkin":{"address":"jaeger-collector.istio-system.svc.cluster.local:9411"}}} and in mesh configmap is {"tracing":{"zipkin":{"address":"jaeger-collector.istio-system.svc.cluster.local:14268"}}}). This often happens when re-installing Istio control-plane and updating the mesh configmap.
Warning [IST0149] (Pod sleep-557747455f-xnjt9.default) The environment variable configured with the running proxies on pods does not match the environment variable defined in the mesh configmap (dismatch environment variable in pod is {"tracing":{"zipkin":{"address":"jaeger-collector.istio-system.svc.cluster.local:9411"}}} and in mesh configmap is {"tracing":{"zipkin":{"address":"jaeger-collector.istio-system.svc.cluster.local:14268"}}}). This often happens when re-installing Istio control-plane and updating the mesh configmap.
......

```

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 1]
```diff
@@ -535,3 +535,15 @@ messages:
     description: "user namespace should be injectable if Istio is installed with enableNamespacesByDefault enabled and neither injection label is set."
     template: "is enabled for Istio injection, as Istio is installed with enableNamespacesByDefault as true."
     url: "https://istio.io/latest/docs/reference/config/analysis/ist0148/"
+
+  - name: "IstioProxyConfigMismatch"
+    code: IST0149
+    level: Warning
+    description: "The environment variable configured with the running proxies on pods does not match the environment variable defined in the mesh configmap."
+    template: "The environment variable configured with the running proxies on pods does not match the environment variable defined in the mesh configmap (dismatch environment variable in pod is %s and in mesh configmap is %s). This often happens when re-installing Istio control-plane and updating the mesh configmap."
```
Comment (shamsher31) [inline galley/pkg/config/analysis/msg/messages.yaml:543]:
```suggestion
    template: "The environment variable %q configured with the running proxies on pods does not match the environment variable %q defined in the mesh configmap. This often happens when re-installing the Istio control plane and updating the mesh configmap."
```

Comment (zhlsunshine) [reply galley/pkg/config/analysis/msg/messages.yaml]: It's done, thanks @shamsher31!  And below is the output:

```
......
Warning [IST0149] (Pod httpbin-74fb669cc6-9tcrx.default) The environment variable "{\"tracing\":{\"zipkin\":{\"address\":\"jaeger-collector.istio-system.svc.cluster.local:9411\"}}}" configured with the running proxies on pods does not match the environment variable "{\"tracing\":{\"zipkin\":{\"address\":\"jaeger-collector.istio-system.svc.cluster.local:14268\"}}}" defined in the mesh configmap. This often happens when re-installing the Istio control plane and updating the mesh configmap.
Warning [IST0149] (Pod sleep-557747455f-xnjt9.default) The environment variable "{\"tracing\":{\"zipkin\":{\"address\":\"jaeger-collector.istio-system.svc.cluster.local:9411\"}}}" configured with the running proxies on pods does not match the environment variable "{\"tracing\":{\"zipkin\":{\"address\":\"jaeger-collector.istio-system.svc.cluster.local:14268\"}}}" defined in the mesh configmap. This often happens when re-installing the Istio control plane and updating the mesh configmap.
......
```

**Review by howardjohn — CHANGES_REQUESTED** [review]
This doesn't take into account a large number of factors that will lead to excessive false positives. The source of the env var can come from MANY different areas - multiple configmaps, filemounts, explicitly setting it on the pod, annotations, etc. Revision can also be set on a per-pod basis, not just at the namespace level. This code assumes the happy path of a single configmap and no customization which is not sufficient; especially when this is reported for every pod this will lead to a lot of noise which we are already seeing is a big issue for analyzer.

I would be hesitant to move forward with this at all as its quite a mess to sort out, but if you are are still interested I think we should make sure we handle all of these cases... but if we handle each one the code will be incredibly complex.

File: galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		}
+		return true
+	})
+
+	c.ForEach(collections.K8SCoreV1Pods.Name(), func(r *resource.Instance) bool {
+		pod := r.Message.(*v1.Pod)
+		if nsRevision, ok := injectedNamespaces[pod.GetNamespace()]; ok {
+			meshCMName := util.GetMeshConfigMapName(nsRevision)
+			if cm, cmOK := proxyConfigMap[meshCMName]; cmOK {
+				proxyConfig, err := loadMeshConfig(cm)
```
[the remark this answers is not in the document]
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go:87]: thanks, @howardjohn , I will refactor these code.

File: galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		DefaultConfig: &meshconfig.ProxyConfig{
+			ProxyMetadata: map[string]string{},
+		},
+	}
+	if err := gogoprotomarshal.ApplyYAML(meshCM.Data[util.MeshConfig], meshConfig); err != nil {
+		return nil, err
+	}
+	// Only support to validate environment variable in ProxyMetadata and PROXY_CONFIG
+	proxyConfig := meshConfig.DefaultConfig.ProxyMetadata
+	bTracing, err := json.Marshal(meshConfig.DefaultConfig.GetTracing())
```
[the remark this answers is not in the document]
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go:127]: @howardjohn good idea,  I will change it.

File: galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go]: Hi @howardjohn ,  done!

File: galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/injection/injection-proxyconfig.go]: Hi @howardjohn, done!

**Comment by zhlsunshine** [comment]
> This doesn't take into account a large number of factors that will lead to excessive false positives. The source of the env var can come from MANY different areas - multiple configmaps, filemounts, explicitly setting it on the pod, annotations, etc. Revision can also be set on a per-pod basis, not just at the namespace level. This code assumes the happy path of a single configmap and no customization which is not sufficient; especially when this is reported for every pod this will lead to a lot of noise which we are already seeing is a big issue for analyzer.
> 
> I would be hesitant to move forward with this at all as its quite a mess to sort out, but if you are are still interested I think we should make sure we handle all of these cases... but if we handle each one the code will be incredibly complex.

Thanks @howardjohn, I understand your worries,  it should be a problem if customized environment variable will take precedence.  Therefore, I will temporarily close this PR.

**Pull request closed without merging**